File: julia/src/STM.jl
# File: STM.jl
# Project: src
# Created Date: 14/06/2022
# Author: Shun Suzuki
# -----
# Last Modified: 14/06/2022
# Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
# -----
# Copyright (c) 2022 Shun Suzuki. All rights reserved.
# 

mutable struct STM
    _ptr::Ptr{Cvoid}
    get_frequency
    set_frequency
    get_sampling_frequency_division
    set_sampling_frequency_division
    get_sampling_frequency
    function STM(ptr::Ptr{Cvoid})
        m = new(ptr)
        m.get_frequency = () -> autd3capi.autdstm_frequency(m._ptr)
        m.set_frequency = (freq) -> autd3capi.autdstm_set_frequency(m._ptr, Float64(freq))
        m.get_sampling_frequency_division = () -> autd3capi.autdstm_sampling_frequency_division(m._ptr)
        m.set_sampling_frequency_division = (freq_div) -> autd3capi.autdstm_set_sampling_frequency_division(m._ptr, UInt32(freq_div))
        m.get_sampling_frequency = () -> autd3capi.autdstm_sampling_frequency(m._ptr)
        finalizer(m -> autd3capi.autd_delete_stm(m._ptr), m)
        m
    end
end

File: matlab/autd3/STM.m
%{
%File: STM.m
%Project: autd3
%Created Date: 10/06/2022
%Author: Shun Suzuki
%-----
%Last Modified: 11/06/2022
%Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
%-----
%Copyright (c) 2022 Shun Suzuki. All rights reserved.
%
%}

classdef STM < Body

    properties
        freqeuncy
        sampling_frequency_division
    end

    methods

        function obj = STM()
            obj = obj@Body();
        end

        function set.freqeuncy(obj, value)
            obj.freqeuncy = value;
            calllib('autd3capi', 'AUTDSTMSetFrequency', obj.ptr, value);
        end

        function freq = get.freqeuncy(obj)
            freq = calllib('autd3capi', 'AUTDSTMFrequency', obj.ptr);
        end

        function set.sampling_frequency_division(obj, value)
            obj.sampling_frequency_division = value;
            calllib('autd3capi', 'AUTDSTMSetSamplingFrequencyDivision', obj.ptr, value);
        end

        function div = get.sampling_frequency_division(obj)
            div = calllib('autd3capi', 'AUTDSTMSamplingFrequencyDivision', obj.ptr);
        end

        function freq = sampling_frequency(obj)
            freq = calllib('autd3capi', 'AUTDSTMSamplingFrequency', obj.ptr);
        end

        function delete(obj)

            if obj.ptr.Value ~= 0
                calllib('autd3capi', 'AUTDDeleteSTM', obj.ptr);
                obj.ptr = libpointer('voidPtr', 0);
            end

        end

    end

end
